test/astarte_vmq_plugin_connection_synchronizer_test.exs

```diff
@@ -0,0 +1,165 @@
+  end
+
+  setup %{chan: chan} do
+    test_pid = self()
+
+    {:ok, consumer_tag} =
+      Queue.subscribe(chan, @queue_name, fn payload, meta ->
+        send(test_pid, {:amqp_msg, payload, meta})
+      end)
+
+    on_exit(fn ->
+      Queue.unsubscribe(chan, consumer_tag)
+    end)
+
+    :ok
+  end
+
+  test "Standard connection flow" do
+    connection_timestamp = now_us_x10_timestamp()
+
+    # no reconciler
+    assert Registry.lookup(AstarteVMQPluginConnectionSynchronizer.Registry, @device_base_path) ==
+             []
+
+    # start the reconciler
+    reconciler_pid = get_connection_reconciler_process!(@device_base_path)
+
+    assert Registry.lookup(AstarteVMQPluginConnectionSynchronizer.Registry, @device_base_path) ==
+             [
+               {reconciler_pid, nil}
+             ]
+
+    # connect the device
+    Synchronizer.handle_connection(reconciler_pid, connection_timestamp)
+
+    # Make sure messages arrived
+    Process.sleep(100)
+
+    # the reconciler should be dead by now
+    assert Registry.lookup(AstarteVMQPluginConnectionSynchronizer.Registry, @device_base_path) ==
+             []
+
+    assert_receive {:amqp_msg, "",
+                    %{headers: connection_headers, timestamp: ^connection_timestamp}}
+
+    assert %{"x_astarte_msg_type" => "connection"} = amqp_headers_to_map(connection_headers)
+  end
+
+  test "Disconnection and reconnection swapped, but near in time, are correctly reordered" do
+    connection_timestamp = now_us_x10_timestamp()
+    disconnection_timestamp = connection_timestamp + 1
+    reconnection_timestamp = connection_timestamp + 2
+
+    # no reconciler
+    assert Registry.lookup(AstarteVMQPluginConnectionSynchronizer.Registry, @device_base_path) ==
+             []
+
+    # start the reconciler
+    reconciler_pid = get_connection_reconciler_process!(@device_base_path)
+
+    assert Registry.lookup(AstarteVMQPluginConnectionSynchronizer.Registry, @device_base_path) ==
+             [
+               {reconciler_pid, nil}
+             ]
+
+    # connect the device
+    Synchronizer.handle_connection(reconciler_pid, connection_timestamp)
+
+    # Make sure messages arrived
+    Process.sleep(100)
+
+    # the reconciler should be dead by now
+    assert Registry.lookup(AstarteVMQPluginConnectionSynchronizer.Registry, @device_base_path) ==
+             []
+
+    assert_receive {:amqp_msg, "",
+                    %{headers: connection_headers, timestamp: ^connection_timestamp}}
+
+    assert %{"x_astarte_msg_type" => "connection"} = amqp_headers_to_map(connection_headers)
+
+    # Now, another reconciler comes in hand
+    another_reconciler_pid = get_connection_reconciler_process!(@device_base_path)
+
+    # make some strange things à la VMQ, like a connection just before a disconnection
+    Synchronizer.handle_connection(another_reconciler_pid, reconnection_timestamp)
+
+    Synchronizer.handle_disconnection(another_reconciler_pid, disconnection_timestamp)
+
```

Comment:
Ideally, you should test this executing one of the two in a different process. If you recall, the previous version was working correctly in tests but failing in a real deployment, and the reason was exactly this. We should mirror this in tests to avoid reintroducing back some regression that might break the code in the same way (e.g. unintentionally linking processes which then die)